### PR TITLE
feat: fusdc liquidity pool borrow and repay facets

### PR DESCRIPTION
**Pull request by 0xpatrickdev**
refs: #10390

## Description
- Adds `borrower` and `repayer` facets to the `Liquidity Pool`
- Track `outstandingLends: Amount` in state for `checkPoolBalance` invariant
- Track `PoolMetrics` like cumulative borrows, repays, and fees

### Security Considerations
Handles payment allocations, but is able to do so synchronously using a temporary seats provided by the callers.

### Scaling Considerations
Makes a vstorage writes for each borrow and repay.

### Documentation Considerations
Updates code comments for maintainers.

### Testing Considerations
Tests use test-only methods on the createFacet to simulate fees from a borrow/repay sequence. Tests ensure `zcf.atomicRearrange()` in `.borrow()` and `.repay()` will not fail and trigger `zcf.shutdownWithFailure()`.

### Upgrade Considerations
None, unreleased

### CODE REVIEW

File: packages/fast-usdc/src/exos/advancer.js [hunk 1]
```diff
@@ -112,8 +106,9 @@ export const prepareAdvancerKit = (
         async handleTransactionEvent(evidence) {
           await null;
           try {
-            // TODO poolAccount might be a vow we need to unwrap
-            const { assetManagerFacet, poolAccount } = this.state;
+            const { assetManagerFacet } = this.state;
+            // TODO is there a better way than this?
+            const poolAccount = await when(this.state.poolAccount);
```
Comment (dckc) [inline packages/fast-usdc/src/exos/advancer.js]: make the caller pass an `OrchestrationAccount`, not a `Vow<OrchestrationAccount>`

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 2]
```diff
@@ -89,60 +102,145 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
...
+            amount,
+            available,
+            outstandingLends,
+            poolStats,
+          );
+
+          // XXX COMMIT POINT?
+          const { USDC: paymentP } = await withdrawFromSeat(zcf, poolSeat, {
+            USDC: amount,
+          });
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]: as discussed, rather than dealing with payments here, let's have the caller supply a seat.
The advancer can make a temp seat.

Then I think we can avoid turn boundaries in `borrow` and `repay`.

File: packages/fast-usdc/src/exos/advancer.js (hunk 1 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/advancer.js]: This is no longer part of the PR

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 3]
```diff
@@ -89,60 +111,154 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
...
+      borrower: {
+        lookupBalance() {
+          const { poolSeat } = this.state;
+          return poolSeat.getAmountAllocated('USDC', usdc.brand);
+        },
         /**
-         * @param {Amount<'nat'>} amount
-         * @param {Payment<'nat'>} payment
+         * @param {ZCFSeat} assetManagerSeat
+         * @param {{ USDC: Amount<'nat'>}} amounts
```
Comment (0xpatrickdev) [inline packages/fast-usdc/src/exos/liquidity-pool.js]:
```suggestion
         * @param {{ USDC: Amount<'nat'>}} amountKWR
```

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 4]
```diff
@@ -45,31 +44,54 @@ const makeDust = brand => AmountMath.make(brand, 1n);
...
  * @param {Brand} USDC
+ * @param {Amount<'nat'>} outstandingLends
  */
-const checkPoolBalance = (poolSeat, shareWorth, USDC) => {
+const checkPoolBalance = (poolSeat, shareWorth, USDC, outstandingLends) => {
   const available = poolSeat.getAmountAllocated('USDC', USDC);
   const dust = makeDust(USDC);
-  isEqual(add(available, dust), shareWorth.numerator) ||
+  const virtualTotal = add(add(available, dust), outstandingLends);
+  isEqual(virtualTotal, shareWorth.numerator) ||
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]:
```suggestion
  const encumberedBalance = add(add(available, dust), outstandingLends);
  isEqual(encumberedBalance, shareWorth.numerator) ||
```

Comment (dckc) [reply packages/fast-usdc/src/exos/liquidity-pool.js]: `checkPoolBalance` docstring needs updating

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 5]
```diff
@@ -45,31 +44,54 @@ const makeDust = brand => AmountMath.make(brand, 1n);
...
     Fail`🚨 pool balance ${q(available)} inconsistent with shareWorth ${q(shareWorth)}`;
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]: I wonder if this should crash the vat. (out of scope of this PR, I suppose)

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 6]
```diff
@@ -45,31 +44,54 @@ const makeDust = brand => AmountMath.make(brand, 1n);
...
+ * }} RepayPaymentKWR
+ */
+
 /**
  * @param {Zone} zone
- * @param {ZCF} zcf
- * @param {Brand<'nat'>} USDC
+ * @param {object} powers
+ * @param {ZCF} powers.zcf
+ * @param {{brand: Brand<'nat'>; issuer: Issuer<'nat'>;}} powers.usdc
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]: I don't see any use of `usdc.issuer`. Why this change?

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 7]
```diff
@@ -89,60 +111,154 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
...
-      feeSink: {
+      borrower: {
+        lookupBalance() {
+          const { poolSeat } = this.state;
+          return poolSeat.getAmountAllocated('USDC', usdc.brand);
+        },
         /**
-         * @param {Amount<'nat'>} amount
-         * @param {Payment<'nat'>} payment
+         * @param {ZCFSeat} assetManagerSeat
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]: `assetManagerSeat` seems like a left-over.

```suggestion
         * @param {ZCFSeat} toSeat
```

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 8]
```diff
@@ -89,60 +111,154 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
...
+      const poolStats = /** @type {PoolStats} */ harden({
+        totalBorrows: makeEmpty(usdc.brand),
+        totalContractFees: makeEmpty(usdc.brand),
+        totalPoolFees: makeEmpty(usdc.brand),
+        totalRepays: makeEmpty(usdc.brand),
+      });
       return {
-        shareMint,
-        shareWorth,
+        contractSeat,
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]: It seems like there are lots of seats in this contract, so the name `contractSeat` doesn't help much. I had to look at the rest of the code to see what this one is for.

We've used `feeSeat` in other contracts. There's more than one kind of fee here, but I think I still prefer it.

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 9]
```diff
@@ -89,60 +111,154 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
...
+          this.facets.external.publishPoolMetrics();
         },
+        // TODO repay failed `LOA.deposit()`
       },
+      repayer: {
+        /**
+         * @param {ZCFSeat} assetManagerSeat
+         * @param {RepayAmountKWR} amounts
+         */
+        repay(assetManagerSeat, amounts) {
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]:
```suggestion
        repay(fromSeat, amounts) {
```

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 10]
```diff
@@ -45,31 +44,54 @@ const makeDust = brand => AmountMath.make(brand, 1n);
...
-export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
+export const prepareLiquidityPoolKit = (zone, { zcf, usdc, tools }) => {
   return zone.exoClassKit(
     'Liquidity Pool',
     {
-      feeSink: M.interface('feeSink', {
-        receive: M.call(AmountShape, PaymentShape).returns(M.promise()),
+      borrower: M.interface('borrower', {
+        lookupBalance: M.call().returns(AmountShape),
+        borrow: M.call(SeatShape, AmountKeywordRecordShape).returns(),
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]: The method implementation assumes the keywords are right but this pattern doesn't ensure it.

Since the USDC brand is in scope, the pattern can constrain that too.

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 11]
```diff
@@ -89,60 +111,154 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
...
+
+          // COMMIT POINT
+          try {
+            zcf.atomicRearrange(
+              harden([
+                [
+                  assetManagerSeat,
+                  poolSeat,
+                  rest,
+                  { USDC: add(amounts.PoolFee, amounts.Principal) },
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js:223]: Add a test where the `fromSeat` doesn't have enough to cover these amounts, please? I think the `atomicRearrange` will fail, taking the contract down with it.

It looks like we should push this `add()` down into `repayCalc`. And `repayCalc` should get the seat allocation and the amounts. This isn't a zoe offer handler, so we shouldn't rely on the caller to ensure that the seat has enough.

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 12]
```diff
@@ -89,60 +111,154 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
...
-          const { recorder } = this.state.shareWorthRecorderKit;
+        publishPoolMetrics() {
+          const { poolStats, shareWorth, poolSeat } = this.state;
+          const { recorder } = this.state.poolMetricsRecorderKit;
           // Consumers of this .write() are off-chain / outside the VM.
           // And there's no way to recover from a failed write.
           // So don't await.
-          void recorder.write(shareWorth);
+          void recorder.write(
+            /** @type {PoolMetrics} */ ({
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]: Why the cast here?

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 13]
```diff
@@ -89,60 +111,154 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
...
+              availableBalance: poolSeat.getAmountAllocated('USDC', usdc.brand),
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]: I wonder if that's a slight misnomer. If the poolSeat allocation is 1 USDC, you can't borrow all of it. The poolSeat can never go empty.

File: packages/fast-usdc/test/fast-usdc.contract.test.ts [hunk 14]
```diff
@@ -213,12 +219,32 @@ test('LP deposits, earns fees, withdraws', async t => {
...
+      amt.USDC,
+      splits.Principal,
+      'testBorrow returns requested amount',
+    );
+
+    const repayPayments = {};
+    for (const [kw, amount] of Object.entries(splits)) {
+      const pmt = await utils.pourPayment(amount);
+      repayPayments[kw] = pmt;
+    }
```
Comment (dckc) [inline packages/fast-usdc/test/fast-usdc.contract.test.ts]: This should avoid the `@ts-expect-error` below:

```suggestion
    const repayPayments = await deeplyFulfilledObject(objectMap(splits, utils.pourPayment));
```

**Review by dckc — COMMENTED** [review]
this is before the fixups

File: packages/fast-usdc/src/exos/liquidity-pool.js (hunk 13 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/liquidity-pool.js]: Per our convo, we'll publish `encumberedBalance` in `PoolMetrics` instead of `availableBalance`

File: packages/fast-usdc/src/exos/liquidity-pool.js (hunk 12 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/liquidity-pool.js]: removed

File: packages/fast-usdc/src/exos/liquidity-pool.js (hunk 11 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/liquidity-pool.js:223]: Please see "repay fails when amounts do not match seat allocation" and "repay fails when seat allocation does not equal amounts". 

I did not push the `add()` into `repayCalc`, but believe we can confidence `atomicRearrange` will not fail. There are additional tests that show the AKWR guards working as expected.

File: packages/fast-usdc/src/exos/liquidity-pool.js (hunk 10 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/liquidity-pool.js]: Good suggestion, done

File: packages/fast-usdc/src/exos/liquidity-pool.js (hunk 6 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/liquidity-pool.js]: vestigial, removed

File: packages/fast-usdc/src/exos/liquidity-pool.js (hunk 4 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/exos/liquidity-pool.js]: Since encumbered and unencumbered balance are mutually exclusive, I don't think `encumberedBalance` is a good name here.

I did rename `outstandingLends` to `encumberedBalance` and use `unencumberedBalance` to represent the current poolSeat allocation. I'm calling the sum of encumbered and unencumbered `grossTotal`, but am open to additional suggestions.

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 15]
```diff
@@ -89,60 +121,149 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
      */
     (shareMint, node) => {
       const { brand: PoolShares } = shareMint.getIssuerRecord();
-      const proposalShapes = makeProposalShapes({ USDC, PoolShares });
+      const proposalShapes = makeProposalShapes({
+        USDC,
+        PoolShares,
+      });
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]: nit: superfluous whitespace change?

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 16]
```diff
@@ -89,60 +121,149 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
...
       const shareWorth = makeParity(makeDust(USDC), PoolShares);
       const { zcfSeat: poolSeat } = zcf.makeEmptySeatKit();
-      const shareWorthRecorderKit = tools.makeRecorderKit(node, RatioShape);
+      const { zcfSeat: feeSeat } = zcf.makeEmptySeatKit();
+      const poolMetricsRecorderKit = tools.makeRecorderKit(
+        node,
+        PoolMetricsShape,
+      );
+      /** used for `checkPoolBalance` invariant */
+      const encumberedBalance = makeEmpty(USDC);
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js]: Perhaps expand to say: poolSeat USDC allocation plus outstanding borrows.

I wonder if this docstring would get associated with `state.encumberedBalance` if it were moved down to the `return` statement.

File: packages/fast-usdc/src/pool-share-math.js [hunk 17]
```diff
@@ -120,3 +125,66 @@ export const withFees = (shareWorth, fees) => {
...
   return makeRatioFromAmounts(balancePost, shareWorth.denominator);
 };
+
+/**
+ *
+ * @param {Amount<'nat'>} requested
+ * @param {Amount<'nat'>} poolSeatAllocation
+ * @param {Amount<'nat'>} encumberedBalance
+ * @param {PoolStats} poolStats
+ * @throws {Error} if conditions are not met
```
Comment (dckc) [inline packages/fast-usdc/src/pool-share-math.js]: Let's either say what conditions or skip this `@throws`.

```suggestion
 * @throws {Error} if too much is requested
```

File: packages/fast-usdc/src/pool-share-math.js [hunk 18]
```diff
@@ -120,3 +125,66 @@ export const withFees = (shareWorth, fees) => {
...
+  encumberedBalance,
+  poolStats,
+) => {
+  (isEqual(fromSeatAllocation.Principal, amounts.Principal) &&
+    isEqual(fromSeatAllocation.PoolFee, amounts.PoolFee) &&
+    isEqual(fromSeatAllocation.ContractFee, amounts.ContractFee)) ||
+    Fail`Cannot repay. From seat allocation ${q(fromSeatAllocation)} does not equal amounts ${q(amounts)}.`;
+
+  !isGTE(amounts.Principal, encumberedBalance) ||
+    isEqual(amounts.Principal, encumberedBalance) ||
```
Comment (dckc) [inline packages/fast-usdc/src/pool-share-math.js]: why test both >= and = except to give separate errors?

if x = y, then x >= y is true, and !(x >= y) will fail. The `isEqual` branch can never be reached.

File: packages/fast-usdc/src/pool-share-math.js [hunk 19]
```diff
@@ -120,3 +125,66 @@ export const withFees = (shareWorth, fees) => {
...
+    Fail`Cannot repay. Principal ${q(amounts.Principal)} exceeds encumberedBalance ${q(encumberedBalance)}.`;
+
+  return harden({
+    shareWorth: withFees(shareWorth, amounts.PoolFee),
```
Comment (dckc) [inline packages/fast-usdc/src/pool-share-math.js:180]: maybe `withFees` should be inlined... or at least not exported any more.

File: packages/fast-usdc/test/fast-usdc.contract.test.ts [hunk 20]
```diff
@@ -205,23 +220,310 @@ test('LP deposits, earns fees, withdraws', async t => {
...
+      // @ts-expect-error intentionally incorrect KW
+      { Fee: usdc.make(1n) },
+    ),
+    {
+      message: /Must have missing properties \["USDC"\]/,
+    },
+  );
+
+  // LPs can still withdraw (contract did not shutdown)
+  await lps.alice.withdraw(0.5);
```
Comment (dckc) [inline packages/fast-usdc/test/fast-usdc.contract.test.ts:342]: :clap:

File: packages/fast-usdc/test/fast-usdc.contract.test.ts [hunk 21]
```diff
@@ -205,23 +220,310 @@ test('LP deposits, earns fees, withdraws', async t => {
...
+  for (const r of Object.values(repayResult)) {
+    t.is(r.value, 0n, 'testRepay consumes all payments');
+  }
+
+  await eventLoopIteration();
+  t.like(await E(subscriber).getUpdateSince(), {
+    value: {
+      encumberedBalance: {
+        value: 0n,
+      },
```
Comment (dckc) [inline packages/fast-usdc/test/fast-usdc.contract.test.ts:502]: how did the `encumberedBalance` get to zero? and if it's zero, how can alice withdraw below?

File: packages/fast-usdc/src/exos/liquidity-pool.js [hunk 22]
```diff
@@ -1,58 +1,76 @@
...
 
 /**
- * Use of pool-share-math in offer handlers below assumes that
- * the pool balance represented by the USDC allocation in poolSeat
- * is the same as the pool balance represented by the numerator
- * of shareWorth.
+ * Verifies that the total pool balance (unencumbered + encumbered) matches the
+ * shareWorth numerator. The total pool balance consists of:
+ * 1. unencumbered balance - USDC available in the pool for borrowing
+ * 2. encumbered balance - USDC currently lent out
```
Comment (dckc) [inline packages/fast-usdc/src/exos/liquidity-pool.js:40]: isn't the encumbered balance the pool seat allocation minus outstanding borrows?

**Review by dckc — COMMENTED** [review]
isn't the encumbered balance the pool seat allocation minus outstanding borrows?

File: packages/fast-usdc/src/exos/liquidity-pool.js (hunk 22 above)
Comment (dckc) [reply packages/fast-usdc/src/exos/liquidity-pool.js:40]: ok. after discussion... the docs here are right.

File: packages/fast-usdc/src/exos/liquidity-pool.js (hunk 16 above)
Comment (dckc) [reply packages/fast-usdc/src/exos/liquidity-pool.js]: Perhaps expand to say: ~~poolSeat USDC allocation plus outstanding borrows.~~ aka outstanding borrows

File: packages/fast-usdc/src/pool-share-math.js (hunk 18 above)
Comment (dckc) [reply packages/fast-usdc/src/pool-share-math.js]: > The isEqual branch can never be reached.

oops. this is an or.

but `(!(x >= y) || x = y)` is just `y >= x`, yes? so we can shorten this to `isGTE(encumberedBalance, amounts.Principal)`

**Review by dckc — APPROVED** [review]
some minor comments / suggestions

Sorry some of them were based on confusion.

strongest suggestion is to rewrite `(!(x >= y) || x = y)` as `y >= x` in pool-share-math

File: packages/fast-usdc/src/pool-share-math.js (hunk 18 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/pool-share-math.js]: Thanks, amended to simplify. I think soon™️ I'll have the hang of `AmountMath`. Looking at the source, it's interesting to see virtually every comparator is implemented with `isGTE`.

File: packages/fast-usdc/src/pool-share-math.js (hunk 19 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/src/pool-share-math.js:180]: Agree, but there's an existing unit test that relies on it so I've left the export for now

File: packages/fast-usdc/test/fast-usdc.contract.test.ts (hunk 21 above)
Comment (0xpatrickdev) [reply packages/fast-usdc/test/fast-usdc.contract.test.ts:502]: We expect `0n` here since all borrows have been returned
